### PR TITLE
Correct context uses

### PR DESCRIPTION
**Pull request by MakeNowJust**
`doAPI`'s `context` does not work due to `WithContext` immutability. `Authenticate` and `RegisterApp` does not use `context` although they needs `context`, so their `context` does not work also. This pull request fixes them and adds test cases.

### CODE REVIEW

File: streaming.go [hunk 1]
```diff
@@ -89,6 +89,7 @@ func (c *Client) streaming(ctx context.Context, p string, params url.Values) (ch
 			}
 			req, err := http.NewRequest(http.MethodGet, u.String(), in)
 			if err == nil {
+				req = req.WithContext(ctx)
```
Comment (MakeNowJust) [inline streaming.go]: I don't know that this line is really needed.

File: apps_test.go [hunk 2]
```diff
@@ -41,3 +42,25 @@ func TestRegisterApp(t *testing.T) {
...
+func TestRegisterAppWithCancel(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(3 * time.Second)
+		fmt.Fprintln(w, `{"client_id": "foo", "client_secret": "bar"}`)
+		return
+	}))
+	defer ts.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go cancel()
```
Comment (mattn) [inline apps_test.go]: I don't make sure but test seems racy.

File: mastodon_test.go [hunk 3]
```diff
@@ -42,6 +43,29 @@ func TestAuthenticate(t *testing.T) {
...
+	}))
+	defer ts.Close()
+
+	client := NewClient(&Config{
+		Server:       ts.URL,
+		ClientID:     "foo",
+		ClientSecret: "bar",
+	})
+	ctx, cancel := context.WithCancel(context.Background())
+	go cancel()
```
Comment (mattn) [inline mastodon_test.go]: ditto

File: mastodon_test.go [hunk 4]
```diff
@@ -79,6 +103,31 @@ func TestPostStatus(t *testing.T) {
...
```
Comment (mattn) [inline mastodon_test.go]: ditto

File: apps_test.go (hunk 2 above)
Comment (MakeNowJust) [inline apps_test.go]: I will remove `go` keyword. I thought it is blocking function.

File: streaming.go (hunk 1 above)
Comment (mattn) [inline streaming.go]: yes, required. there are two chance to handle cancel in this part. the first is while sending request. second is while reading. the second already implemented but first is not. so please remain your code.

**Comment by mattn** [comment]
LGTM. Thanks!

**Pull request merged**